File: backend/app/api/v1/endpoints/faq.py
"""
FAQ API endpoints
"""
# Standard library imports
from typing import List

# Third-party imports
from fastapi import APIRouter, Depends, Query

# Local imports
from app.core.deps import get_admin_user, get_current_user
from app.models.user import User
from app.schemas.faq import (
    FAQBulkAction,
    FAQCreate,
    FAQSearchQuery,
    FAQUpdate,
    FAQVoteRequest
)
from app.schemas.base import StandardResponse
from app.services.faq_service import faq_service

router = APIRouter()


@router.get("", response_model=StandardResponse[dict])
async def get_faqs(
    q: str = Query(None, description="Search query"),
    category: str = Query(None, description="Filter by category"),
    tags: List[str] = Query(None, description="Filter by tags"),
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    sort_by: str = Query("priority", pattern="^(priority|view_count|created_at)$"),
    sort_order: str = Query("desc", pattern="^(asc|desc)$")
):
    """
    Get all published FAQs with optional search and filters
    """
    query = FAQSearchQuery(
        q=q,
        category=category,
        tags=tags,
        is_published=True,
        page=page,
        per_page=per_page,
        sort_by=sort_by,
        sort_order=sort_order
    )
    
    result = await faq_service.search_faqs(query)
    
    return StandardResponse(
        success=True,
        data=result,
        message="FAQs retrieved successfully"
    )


@router.get("/categories", response_model=StandardResponse[list])
async def get_faq_categories():
    """
    Get FAQ categories with count
    """
    from app.models.faq import FAQ, FAQCategory
    
    categories = []
    for category in FAQCategory:
        count = await FAQ.find({
            "category": category.value,
            "is_published": True
        }).count()
        
        categories.append({
            "value": category.value,
            "label": category.value.replace("_", " ").title(),
            "count": count
        })
    
    return StandardResponse(
        success=True,
        data=categories,
        message="Categories retrieved successfully"
    )


@router.get("/popular", response_model=StandardResponse[dict])
async def get_popular_faqs(
    limit: int = Query(10, ge=1, le=50)
):
    """
    Get most viewed FAQs
    """
    faqs = await faq_service.get_popular_faqs(limit)
    
    return StandardResponse(
        success=True,
        data={
            "items": faqs,
            "total": len(faqs),
            "page": 1,
            "per_page": limit
        },
        message="Popular FAQs retrieved successfully"
    )


@router.get("/search", response_model=StandardResponse[dict])
async def search_faqs(
    query_params: FAQSearchQuery = Depends()
):
    """
    Search FAQs with advanced filters
    """
    result = await faq_service.search_faqs(query_params)
    
    return StandardResponse(
        success=True,
        data=result,
        message="Search results retrieved successfully"
    )


@router.post("", response_model=StandardResponse[dict])
async def create_faq(
    faq_data: FAQCreate,
    current_user: User = Depends(get_admin_user)
):
    """
    Create a new FAQ (Admin only)
    """
    faq = await faq_service.create_faq(faq_data)
    
    return StandardResponse(
        success=True,
        data=faq,
        message="FAQ created successfully"
    )


@router.get("/{faq_id}", response_model=StandardResponse[dict])
async def get_faq(faq_id: str):
    """
    Get FAQ by ID and increment view count
    """
    faq = await faq_service.get_faq(faq_id)
    
    return StandardResponse(
        success=True,
        data=faq,
        message="FAQ retrieved successfully"
    )


@router.get("/{faq_id}/related", response_model=StandardResponse[dict])
async def get_related_faqs(faq_id: str):
    """
    Get FAQs related to the given FAQ
    """
    faqs = await faq_service.get_related_faqs(faq_id)
    
    return StandardResponse(
        success=True,
        data={
            "items": faqs,
            "total": len(faqs),
            "page": 1,
            "per_page": len(faqs)
        },
        message="Related FAQs retrieved successfully"
    )


@router.put("/{faq_id}", response_model=StandardResponse[dict])
async def update_faq(
    faq_id: str,
    update_data: FAQUpdate,
    current_user: User = Depends(get_admin_user)
):
    """
    Update FAQ (Admin only)
    """
    faq = await faq_service.update_faq(faq_id, update_data)
    
    return StandardResponse(
        success=True,
        data=faq,
        message="FAQ updated successfully"
    )


@router.delete("/{faq_id}", response_model=StandardResponse[dict])
async def delete_faq(
    faq_id: str,
    current_user: User = Depends(get_admin_user)
):
    """
    Delete FAQ (Admin only)
    """
    result = await faq_service.delete_faq(faq_id)
    
    return StandardResponse(
        success=True,
        data=result,
        message="FAQ deleted successfully"
    )


@router.post("/{faq_id}/vote", response_model=StandardResponse[dict])
async def vote_faq(
    faq_id: str,
    vote_data: FAQVoteRequest,
    current_user: User = Depends(get_current_user)
):
    """
    Vote on FAQ helpfulness (authenticated users)
    """
    result = await faq_service.vote_faq(faq_id, vote_data.is_helpful)
    
    return StandardResponse(
        success=True,
        data=result,
        message=result.get("message", "Vote recorded successfully")
    )


@router.post("/bulk-action", response_model=StandardResponse[dict])
async def bulk_action_faqs(
    bulk_data: FAQBulkAction,
    current_user: User = Depends(get_admin_user)
):
    """
    Perform bulk actions on FAQs (Admin only)
    """
    result = await faq_service.bulk_action(bulk_data.faq_ids, bulk_data.action)
    
    return StandardResponse(
        success=True,
        data=result,
        message=result.get("message", "Bulk action completed successfully")
    )

File: frontend/lib/api/faq.ts
/**
 * FAQ API client functions
 * Updated to use dynamic categories instead of hardcoded enum
 */

import { apiClient } from './api-client';
import { StandardResponse } from '@/lib/types/api';
import { 
  FAQ, 
  FAQCreateData, 
  FAQUpdateData, 
  FAQSearchParams, 
  FAQListResponse,
  FAQVoteRequest,
  FAQVoteResponse
} from '@/lib/types/faq';

// Re-export types for backward compatibility
export type { 
  FAQ, 
  FAQCreateData, 
  FAQUpdateData, 
  FAQSearchParams, 
  FAQListResponse,
  FAQVoteRequest,
  FAQVoteResponse
} from '@/lib/types/faq';

export interface FAQBulkAction {
  action: 'create' | 'publish' | 'unpublish' | 'delete';
  faq_ids?: string[];  // Required for publish/unpublish/delete
  faqs_data?: FAQCreateData[];  // Required for create
}

// FAQ API functions
export const faqAPI = {
  /**
   * Get FAQs with search and filters
   */
  async getFAQs(params?: FAQSearchParams): Promise<StandardResponse<FAQListResponse>> {
    const queryParams = new URLSearchParams();
    
    if (params) {
      if (params.q) queryParams.append('q', params.q);
      if (params.category) queryParams.append('category', params.category);
      if (params.is_published !== undefined) queryParams.append('is_published', params.is_published.toString());
      if (params.page) queryParams.append('page', params.page.toString());
      if (params.per_page) queryParams.append('per_page', params.per_page.toString());
      if (params.sort_by) queryParams.append('sort_by', params.sort_by);
      if (params.sort_order) queryParams.append('sort_order', params.sort_order);
    }
    
    return apiClient.get<StandardResponse<FAQListResponse>>(`/faq?${queryParams.toString()}`);
  },

  /**
   * Get FAQs for admin management (includes unpublished)
   */
  async getAdminFAQs(params?: FAQSearchParams): Promise<StandardResponse<FAQListResponse>> {
    const queryParams = new URLSearchParams();
    
    if (params) {
      if (params.q) queryParams.append('q', params.q);
      if (params.category) queryParams.append('category', params.category);
      if (params.is_published !== undefined) queryParams.append('is_published', params.is_published.toString());
      if (params.page) queryParams.append('page', params.page.toString());
      if (params.per_page) queryParams.append('per_page', params.per_page.toString());
      if (params.sort_by) queryParams.append('sort_by', params.sort_by);
      if (params.sort_order) queryParams.append('sort_order', params.sort_order);
    }
    
    return apiClient.get<StandardResponse<FAQListResponse>>(`/faq/admin?${queryParams.toString()}`);
  },

  /**
   * Get popular FAQs
   */
  async getPopularFAQs(limit: number = 10): Promise<StandardResponse<FAQListResponse>> {
    return apiClient.get<StandardResponse<FAQListResponse>>(`/faq/popular?limit=${limit}`);
  },

  /**
   * Get a single FAQ by ID
   */
  async getFAQ(id: string): Promise<StandardResponse<FAQ>> {
    return apiClient.get<StandardResponse<FAQ>>(`/faq/${id}`);
  },

  /**
   * Get related FAQs
   */
  async getRelatedFAQs(id: string): Promise<StandardResponse<FAQListResponse>> {
    return apiClient.get<StandardResponse<FAQListResponse>>(`/faq/${id}/related`);
  },

  /**
   * Create a new FAQ (Admin only)
   */
  async createFAQ(data: FAQCreateData): Promise<StandardResponse<FAQ>> {
    const response = await apiClient.post<StandardResponse<FAQ>>('/faq', data, { requireAuth: true });
    return response;
  },

  /**
   * Update a FAQ (Admin only)
   */
  async updateFAQ(id: string, data: FAQUpdateData): Promise<StandardResponse<FAQ>> {
    const response = await apiClient.put<StandardResponse<FAQ>>(`/faq/${id}`, data, { requireAuth: true });
    return response;
  },

  /**
   * Delete a FAQ (Admin only)
   */
  async deleteFAQ(id: string): Promise<StandardResponse<{ message: string }>> {
    const response = await apiClient.delete<StandardResponse<{ message: string }>>(`/faq/${id}`, { requireAuth: true });
    return response;
  },

  /**
   * Vote on FAQ helpfulness
   */
  async voteFAQ(id: string, data: FAQVoteRequest): Promise<StandardResponse<FAQVoteResponse>> {
    const response = await apiClient.post<StandardResponse<FAQVoteResponse>>(`/faq/${id}/vote`, data);
    return response;
  },

  /**
   * Perform bulk action on FAQs (Admin only)
   */
  async bulkAction(data: FAQBulkAction): Promise<StandardResponse<{
    affected_count?: number;
    created_count?: number;
    failed_count?: number;
    created_faqs?: FAQ[];
    errors?: any[];
  }>> {
    const response = await apiClient.post<StandardResponse<{
      affected_count?: number;
      created_count?: number;
      failed_count?: number;
      created_faqs?: FAQ[];
      errors?: any[];
    }>>('/faq/bulk-action', data, { requireAuth: true });
    return response;
  },

  /**
   * Bulk create FAQs (Admin only)
   */
  async bulkCreate(faqs: FAQCreateData[]): Promise<StandardResponse<{
    created_count?: number;
    failed_count?: number;
    created_faqs?: FAQ[];
    errors?: any[];
  }>> {
    return this.bulkAction({
      action: 'create',
      faqs_data: faqs
    });
  },
};

// Export individual functions for React Query hooks
export const getFAQs = faqAPI.getFAQs;
export const getAdminFAQs = faqAPI.getAdminFAQs;
export const createFAQ = faqAPI.createFAQ;
export const updateFAQ = faqAPI.updateFAQ;
export const deleteFAQ = faqAPI.deleteFAQ;
export const voteFAQ = faqAPI.voteFAQ;
export const bulkAction = faqAPI.bulkAction;
export const bulkCreate = faqAPI.bulkCreate;
export const searchFAQs = faqAPI.getFAQs; // Alias for consistency